File: L3/Progs/FermatPrimalityTest.py
# -*- coding: utf-8 -*-
# @Author: nevil
# @Date:   2020-08-03 23:02:22
# @Last Modified by:   nevil
# @Last Modified time: 2020-08-03 23:03:57

#Primality testing program with Fermat's theorem.

import random

def GCD(x, y): #gcd calculator
   while(y): 
       x, y = y, x % y 
   return x 

def getNumber():
    return int(input("Enter an integer for Fermat's primality test: "))

def isPrime(n):
    #testing for the most simple cases like: odd numbers
    if n <= 3:
        return True
    elif n % 2 == 0:
        return False


    #testing for 7 times, should be enought to find any pseudo prime numbers
    for i in range(7):
        a = random.randint(2, n-1)
        print(f"Testing for a = {a}")

        if GCD(a,n) == 1:
            if (a ** (n-1)) % n == 1: continue
            else: return False
        else: return False
    return True


while True:
    if isPrime(getNumber()):
        print("This number is possibly prime\n")
    else:
        print("This number is not prime\n")

File: L3/Progs/FermatPrimalityTest.cpp
/*
* @Author: nevil
* @Date:   2020-07-24 16:03:04
* @Last Modified by:   nevil
* @Last Modified time: 2020-07-24 16:08:26
*/
#include<bits/stdc++.h>
using namespace std;

void swap(long long &a, long long &b) {
    long long temp = a;
    a = b;
    b = temp;
}

long long gcd(long long a, long long m) {
    if (a > m) swap(a, m);
    long long r;
    for (;;) {
        r = m % a;
        if (r == 0) return a;
        m = a;
        a = r;
    };
}

long long power(long long a, long long b, long long m) {
    if (a > m) swap(a, m);
    long long c = 1;
    for (;;) {
        if (b % 2 == 1) c = (c * a) % m;
        b = b / 2;
        if (b == 0) return c;
        a = (a * a) % m;
    };
}

bool fermat(long long m) {
    srand(time(NULL));
    if (m == 1) {
        cout << endl;
        return false;
    };
    for (int i = 0; i < 1000; i++) {
        int cnt = i + 1;
        long long a = rand() % m;
        if (a == 0) a = 1;
        if (gcd(a, m) != 1) {
            cout << "\nFailed at the " << "attempt " << cnt << endl;
            cout << "Found a divisor: " << gcd(a, m) << ", thus ";
            return false;
        };
        if (power(a, m - 1, m) != 1) {
            cout << "\nFailed at the " << cnt << ". attempt. ";
            cout << "Found a Fermat witness: " << a << ", thus ";
            return false;
        };
    };
    cout << endl;
    return true;
}

int main() {

    cout << "Fermat primality test" << endl << endl;
    for (;;) {
        long long n;
        cout << "Please input a natural number (0 to quit): ";
        cin >> n;
        if (n == 0) return 0;
        bool b = fermat(n);
        if (b) cout << n << " is probably a prime." << endl << endl;
        if (!b) cout << n << " is not a prime." << endl << endl;
    };
}
